src/common/associatedToken.ts

```diff
@@ -1,12 +1,76 @@
-import { Provider } from "@project-serum/anchor"
+import { Provider, BN } from "@project-serum/anchor"
 import { TOKEN_PROGRAM_ID } from "@project-serum/serum/lib/token-instructions"
-import { ASSOCIATED_TOKEN_PROGRAM_ID, Token, AccountInfo as TokenAccountInfo, MintInfo } from "@solana/spl-token"
-import { AccountInfo, Connection, PublicKey, Signer, TransactionInstruction, ParsedAccountData } from "@solana/web3.js"
+import { struct, u8 } from "@solana/buffer-layout"
+import {
+  ASSOCIATED_TOKEN_PROGRAM_ID,
+  Token,
+  AccountInfo as TokenAccountInfo,
+  MintInfo,
+  NATIVE_MINT
+} from "@solana/spl-token"
+import {
+  AccountInfo,
+  Connection,
+  PublicKey,
+  Signer,
+  TransactionInstruction,
+  ParsedAccountData,
+  SystemProgram
+} from "@solana/web3.js"
 import { useMemo } from "react"
 import { parseMintAccount, parseTokenAccount } from "./accountParser"
 import { Hooks } from "./hooks"
 import { findDerivedAccount } from "."
 
+/** Instructions defined by the program */
+export enum TokenInstruction {
+  InitializeMint = 0,
+  InitializeAccount = 1,
+  InitializeMultisig = 2,
+  Transfer = 3,
+  Approve = 4,
+  Revoke = 5,
+  SetAuthority = 6,
+  MintTo = 7,
+  Burn = 8,
+  CloseAccount = 9,
+  FreezeAccount = 10,
+  ThawAccount = 11,
+  TransferChecked = 12,
+  ApproveChecked = 13,
+  MintToChecked = 14,
+  BurnChecked = 15,
+  InitializeAccount2 = 16,
+  SyncNative = 17,
+  InitializeAccount3 = 18,
+  InitializeMultisig2 = 19,
+  InitializeMint2 = 20
+}
+/** TODO: docs */
+export interface SyncNativeInstructionData {
+  instruction: TokenInstruction.SyncNative
+}
+
+/** TODO: docs */
+export const syncNativeInstructionData = struct([u8("instruction")])
+
+/**
+ * Construct a SyncNative instruction
+ *
+ * @param account   Native account to sync lamports from
+ * @param programId SPL Token program account
+ *
+ * @return Instruction to add to a transaction
+ */
+export function createSyncNativeInstruction(account: PublicKey, programId = TOKEN_PROGRAM_ID): TransactionInstruction {
```

Comment:
You will want to move this into the AssociatedToken class as well

```suggestion
function withSyncNativeInstruction(instructions: TransactionInstruction[], account: PublicKey, programId = TOKEN_PROGRAM_ID): TransactionInstruction {
```